nncf/data/utils.py

```diff
@@ -0,0 +1,106 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from typing import List
+from typing import Iterator
+from typing import Callable
+
+from nncf.data.types import DataSource
+from nncf.data.types import DataItem
+from nncf.data.types import ModelInput
+from nncf.data.dataloader import DataLoader
+from nncf.data.dataloader import DataLoaderImpl
+
+
+def create_dataloader(data_source: DataSource,
+                      transform_fn: Callable[[DataItem], ModelInput]) -> DataLoader:
+    """
+    Wraps the provided custom data source that is an [iterable](https://docs.python.org/3/glossary.html#term-iterable)
+    python object into the NNCF `DataLoader` object.
+
+    :param data_source: Custom data source that is an iterable python object.
+    :param transform_fn: The method that takes a data item returned per iteration
+        through `DataSource` object and transforms it to the model's expected input
+        that can be used for the model inference.
+    :return: The object that implements the NNCF `DataLoader` interface and
+        wraps custom data source.
+    """
+    # Checks that `data_source` is an iterable python object.
+    try:
+        iter(data_source)
+    except TypeError as exc:
+        raise ValueError('The provided `data_source` object is not an iterable object. Please see '
+                         'https://docs.python.org/3/glossary.html#term-iterable to learn more '
+                         'about iterable object.') from exc
+
+    batch_size = _get_batch_size(data_source)
+
+    return DataLoaderImpl(data_source, transform_fn, batch_size)
+
+
+def _get_batch_size(data_source: DataSource) -> int:
+    """
+    Tries to determine the batch size of the provided custom data source.
+    If it was successful returns the number of elements return per iteration
+    from the `data_source` or `1` otherwise.
+
+    :param data_source: Custom data source that is an iterable python object.
+    :return: The number of elements return per iteration from the `data_source` or `1`.
+    """
+    try:
+        import torch
+    except ImportError:
+        torch = None
+
+    try:
+        import tensorflow
+    except ImportError:
+        tensorflow = None
+
+    batch_size = 1
+
+    if torch is not None and isinstance(data_source, torch.utils.data.DataLoader):
+        batch_size = data_source.batch_size
+
+    if tensorflow is not None and isinstance(data_source, tensorflow.data.Dataset):
```

Comment:
Can we support HuggingFace `Dataset` object here? [This](https://github.com/huggingface/datasets/blob/6be722dbbc212d88eb5e550e56431cdcab180eee/src/datasets/arrow_dataset.py#L646) is the definition.